### ISSUE TITLE
Inconsistent behavior of 'update' and '(association)_attributes=' methods with 'accepts_nested_attributes_for'

### ISSUE THREAD

**Issue by ales-vilchytski** [mongoid/mongoid#3920]
I encounter problems with either `embedded_many` or `embedded_one` and `accepts_nested_attributes_for` on next environment:
- Mongoid 4.0.0
- Rails either 4.1 or 4.2.

*Problem*
`update` method on Mongoid document with `accepts_nested_attributes_for` doesn't delete `_destroy` relations and doesn't call `reject_if` method, but updates fields and create new models when attributes passed as hash.
Methods `(association)_attributes=` work correctly - they update field, delete `_destroy` models and add new ones.

I have created gist demonstrating problem:
https://gist.github.com/ales-vilchytski/b462cd8e2ff607fccd44

*Expected result*
It would be great to have similar behavior for both methods.

*What I've found:*
In short, `accepts_nested_attributes_for` sets `options` only for generated `(association)_attributes=` methods by closuring passed object, but this object isn't saved in any other place.
https://github.com/mongoid/mongoid/blob/v4.0.0/lib/mongoid/attributes/nested.rb#L54

But `update` using `Mongoid::Attributes::Processing` at some moment passes empty hash as `options` (then attribute builders doesn't take into account real `allow_destroy` and `reject_if` options):
https://github.com/mongoid/mongoid/blob/v4.0.0/lib/mongoid/attributes/processing.rb#L139

**Comment by TangMonk** [comment]
same issue

can not delete from `_destroy`

**Comment by estolfo** [comment]
Hi @ales-vilchytski

Thanks for your thorough report. I've tested the code you provided and found that the only method that doesn't work in your gist is #delete_by_update ; however, if you change the key "embeds" to "embeds_attributes", then it works. Can you change the key in each of your methods to embeds_attributes and see if they all succeed?

Thanks
Emily

**Comment by ales-vilchytski** [comment]
Hello @estolfo !

I just updated gist (https://gist.github.com/ales-vilchytski/b462cd8e2ff607fccd44) and checked it as a standalone app (I used to run it from Rails earlier...).
So, you're right - `embeds_attributes` works well in all cases.

But  I want to say, that this approach is not handy, especially in Rails with SPA (e.g. Backbone + Marionette). Front exchanges with server by JSON presentation of model and main problem that `<field>_attributes` keys passed to `update` doesn't comply with usual JSON presentation of models.
For example, in one real app there were up to 5 levels of nesting and I finally needed to write something like that (pseudocode):
```
if params_has_nested_attributes?(attributes)
  document.nested_attributes = attributes
else
 document.update(attributes)
end
```

**Comment by estolfo** [comment]
I see what you're saying. I think the change would be non trivial so we'll keep it in mind but feel free to try implementing it in a PR.
I'm going to close the issue because it's not a bug.
